src/main/webapp/src/app/recommendation-form/recommendation-form.component.ts

```diff
@@ -8,18 +9,25 @@ import { FormBuilder} from '@angular/forms';
 })
 export class RecommendationFormComponent{
 
+  @Output() formSubmit: EventEmitter<void> = new EventEmitter()
+
   // Build preference form
   preferenceForm = this.formBuilder.group({
-    price: "",
-    rating:"",
-    type: ""
+    price: ["", Validators.required],
+    rating: ["", Validators.required],
+    type: ["", Validators.required]
   });
 
-  constructor(private readonly formBuilder: FormBuilder) { }   
+  constructor(private readonly formBuilder: FormBuilder, private hiddenGemService: HiddenGemService) { }
 
   // Submit user's preference form
   onSubmit(): void {
     console.log(this.preferenceForm);
+    this.hiddenGemService.findHiddenGemReccomendation(this.preferenceForm.value)
+      .subscribe(hiddenGems => {
+        this.hiddenGemService.updateTop3Gems(hiddenGems);
```

Comment:
How do findHiddenGemRecommendation and updateTop3Gems work? Could we simplify it and just have one function which process the params given in the form and return the top 3? I am unsure what the role of updateTop3Gems is exactly so I might be wrong

Reply:
I agree with the comment above

Reply:
Done. Updated the files and removed the updateTop3Gems function. Use event emitter to send the output as input directly into the list component. Thanks for all your help ❤️

Reply:
@sehhajdkg Good job on getting it to work with event emitter!! It looks great 😄